go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -276,226 +290,62 @@ func (c *Controller) syncServices(key string) error {
 		return err
 	}
 
-	// If idling enabled and there are no endpoints, we need to move the VIP from the main loadbalancer
-	// ,that has the reject option for backends without endpoints, to the idling loadbalancer, that
-	// generates a needPods event.
-	// if not, we delete the vips from the idling lb and move them to their right place.
-
-	vipProtocols := collectServiceVIPs(service)
-	if svcNeedsIdling(service.Annotations) && !util.HasValidEndpoint(service, endpointSlices) {
-		toRemoveFromNonIdling := sets.NewString()
-		for vipProtocol := range vipProtocols {
-			if c.serviceTracker.getLoadBalancer(name, namespace, vipProtocol) != loadbalancer.IdlingLoadBalancer {
-				toRemoveFromNonIdling.Insert(vipProtocol)
-			}
-			vipsTracked.Delete(vipProtocol)
-		}
-		// addServiceToIdlingBalancer adds the vips to service tracker
-		err = c.addServiceToIdlingBalancer(vipProtocols, service)
-		if err != nil {
-			c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToAddToIdlingBalancer",
-				"Error trying to add to Idling LoadBalancer for Service %s/%s: %v", name, namespace, err)
-			return err
-		}
-		err = deleteVIPsFromNonIdlingOVNBalancers(toRemoveFromNonIdling, name, namespace)
-		if err != nil {
-			c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToDeleteOVNLoadBalancer",
-				"Error trying to delete the OVN LoadBalancer while setting up Idling for Service %s/%s: %v", name, namespace, err)
-			return err
+	// Build the abstract LB configs for this service
+	perNodeConfigs, clusterConfigs := buildServiceLBConfigs(service, endpointSlices)
+	klog.V(5).Infof("Built service %s LB cluster-wide configs %#v", key, clusterConfigs)
+	klog.V(5).Infof("Built service %s LB per-node configs %#v", key, perNodeConfigs)
+
+	// Convert the LB configs in to load-balancer objects
+	nodeInfos := c.nodeTracker.allNodes()
+	clusterLBs := buildClusterLBs(service, clusterConfigs, nodeInfos)
+	perNodeLBs := buildPerNodeLBs(service, perNodeConfigs, nodeInfos)
+	klog.V(3).Infof("Service %s has %d cluster-wide and %d per-node configs, making %d and %d load balancers",
+		key, len(clusterConfigs), len(perNodeConfigs), len(clusterLBs), len(perNodeLBs))
+	lbs := append(clusterLBs, perNodeLBs...)
+
+	// Short-circuit if nothing has changed
+	c.alreadyAppliedLock.Lock()
+	existingLBs, ok := c.alreadyApplied[key]
+	c.alreadyAppliedLock.Unlock()
+	if ok && reflect.DeepEqual(lbs, existingLBs) {
+		klog.V(3).Infof("Skipping no-op change for service %s", key)
+	} else {
+		// Actually apply load-balancers to OVN.
+		//
+		// Note: this may fail if a node was deleted between listing nodes and applying.
+		// If so, this will fail and we will resync.
+		if err := ovnlb.EnsureLBs(util.ExternalIDsForObject(service), lbs); err != nil {
```

Comment:
i dont think this is safe, you need to hold the c.alreadyAppliedLock.Lock() (or even better a lock on the service itself) in order to keep a consistent cache of whats in OVN and what externalLBs are in your cache. Otherwise you could have 2 worker threads applying different OVN configs simultaneously for the same service, and updating the cache in reverse order. 

I'm not sure why you have the alreadyApplied cache as well as the LBCache, isn't it really the same information (what is in OVN)?

Reply:
There's no need to lock per-service, because the workqueue ensures that a given service will only ever be processed once at a time. So, in that sense, we already have a per-service lock.

You're right; alreadyApplied somewhat duplicates the LBCache at this point. Right now, LBCache doesn't check configuration, it only caches existence. But it could be made to fully short-cut.

LBCache sort of works like the libovsdb cache; it's only really used for existence tests rather than short-cutting updates. I tried to mirror that behavior.

Reply:
Agree with you on the first point. Would like to see a single cache used for OVN stuff. I know we will remove the cache anyway once we have bindings, but this PR is a huge change and I think for the sake of others understanding the changes in a short amount of time, using a single cache makes things more obvious.

Reply:
+1 on just using the LBCache here, make's it a bit simpler to understand rather than looking here and in the `loadbalancer `package to figure out how the caches are working.

Reply:
I remember why I didn't do this: getting the option code to be fully reversible (in other words, parsed from ovsdb output) is tricky, and this all goes away soon.

The extra cache isn't hurting anything for now.

So, if you don't mind, I'd like to keep this as-is and focus on getting rid of it all :-).